ethdb/olddb/mapmutation.go

```diff
@@ -0,0 +1,304 @@
+//
+// Common pattern:
+//
+// batch := db.NewBatch()
+// defer batch.Rollback()
+// ... some calculations on `batch`
+// batch.Commit()
+func NewHashBatch(tx kv.RwTx, quit <-chan struct{}, tmpdir string) *mapmutation {
+	clean := func() {}
+	if quit == nil {
+		ch := make(chan struct{})
+		clean = func() { close(ch) }
+		quit = ch
+	}
+	return &mapmutation{
+		db:     tx,
+		puts:   make(map[string]map[string][]byte),
+		quit:   quit,
+		clean:  clean,
+		tmpdir: tmpdir,
+	}
+}
+
+func (m *mapmutation) RwKV() kv.RwDB {
+	if casted, ok := m.db.(ethdb.HasRwKV); ok {
+		return casted.RwKV()
+	}
+	return nil
+}
+
+func (m *mapmutation) getMem(table string, key []byte) ([]byte, bool) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if _, ok := m.puts[table]; !ok {
+		return nil, false
+	}
+	var value []byte
+	var ok bool
+	if value, ok = m.puts[table][string(key)]; !ok {
+		return nil, false
+	}
+	return value, ok
+}
+
+func (m *mapmutation) IncrementSequence(bucket string, amount uint64) (res uint64, err error) {
+	v, ok := m.getMem(kv.Sequence, []byte(bucket))
+	if !ok && m.db != nil {
+		v, err = m.db.GetOne(kv.Sequence, []byte(bucket))
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	var currentV uint64 = 0
+	if len(v) > 0 {
+		currentV = binary.BigEndian.Uint64(v)
+	}
+
+	newVBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(newVBytes, currentV+amount)
+	if err = m.Put(kv.Sequence, []byte(bucket), newVBytes); err != nil {
+		return 0, err
+	}
+
+	return currentV, nil
+}
+func (m *mapmutation) ReadSequence(bucket string) (res uint64, err error) {
+	v, ok := m.getMem(kv.Sequence, []byte(bucket))
+	if !ok && m.db != nil {
+		v, err = m.db.GetOne(kv.Sequence, []byte(bucket))
+		if err != nil {
+			return 0, err
+		}
+	}
+	var currentV uint64 = 0
+	if len(v) > 0 {
+		currentV = binary.BigEndian.Uint64(v)
+	}
+
+	return currentV, nil
+}
+
+// Can only be called from the worker thread
+func (m *mapmutation) GetOne(table string, key []byte) ([]byte, error) {
+	if value, ok := m.getMem(table, key); ok {
+		if value == nil {
+			return nil, nil
+		}
+		return value, nil
+	}
+	if m.db != nil {
+		// TODO: simplify when tx can no longer be parent of mutation
+		value, err := m.db.GetOne(table, key)
+		if err != nil {
+			return nil, err
+		}
+
+		return value, nil
+	}
+	return nil, nil
+}
+
+// Can only be called from the worker thread
+func (m *mapmutation) Get(table string, key []byte) ([]byte, error) {
+	value, err := m.GetOne(table, key)
+	if err != nil {
+		return nil, err
+	}
+
+	if value == nil {
+		return nil, ethdb.ErrKeyNotFound
+	}
+
+	return value, nil
+}
+
+func (m *mapmutation) Last(table string) ([]byte, []byte, error) {
+	c, err := m.db.Cursor(table)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer c.Close()
+	return c.Last()
+}
+
+func (m *mapmutation) hasMem(table string, key []byte) bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if _, ok := m.puts[table]; !ok {
+		return false
+	}
+
+	_, ok := m.puts[table][string(key)]
```

Comment:
There is trick to convert []byte to string without allocation - used by strings.Builder:
```
func ByteSlice2String(bs []byte) string {
	return *(*string)(unsafe.Pointer(&bs))
}
```